docs: SMP servers (#1387)
simplex-chat/simplex-chat: docs/SERVER.md

# Hosting your own SMP Server

## Overview

SMP server is the relay server used to pass messages in SimpleX network. SimpleX Chat apps have preset servers (for mobile apps these are smp8, smp9 and smp10.simplex.im), but you can easily change app configuration to use other servers.

SimpleX clients only determine which server is used to receive the messages, separately for each contact (or group connection with a group member), and these servers are only temporary, as the delivery address can change.

> _Please note_: when you change the servers in the app configuration, it only affects which server will be used for the new contacts, the existing contacts will not automatically move to the new servers, but you can move them manually using BETA feature ["Change receiving address"](../blog/20221108-simplex-chat-v4.2-security-audit-new-website.md#change-your-delivery-address-beta) – it will be automated soon.

Author: ```suggestion
_Please note_: when you change the servers in the app configuration, it only affects which server will be used for the new contacts, the existing contacts will not automatically move to the new servers, but you can move them manually using ["Change receiving address"](../blog/20221108-simplex-chat-v4.2-security-audit-new-website.md#change-your-delivery-address-beta) button in contact/member information pages – it will be automated soon.
```

## Installation

> 0. Firtsly, install `smp-server`:

Author: ```suggestion
0. First, install `smp-server`:
```

   - Manual deployment:
     - [Compiling from source](https://github.com/simplex-chat/simplexmq#using-your-distribution)
     - [Using pre-compiled binaries](https://github.com/simplex-chat/simplexmq#install-binaries)

   - Alternatively, you can deploy `smp-server` using:
     - [Docker container](https://github.com/simplex-chat/simplexmq#using-docker-1)
     - [Linode StackScript](https://github.com/simplex-chat/simplexmq#deploy-smp-server-on-linode)

> Manual installation require some preparations:

Author: ```suggestion
Manual installation requires some preliminary actions:
```

1. Create user and group for `smp-server`:

   ```sh
   sudo useradd -m smp
   ```

2. Create necessary directories and assign permissions:

   ```sh
   sudo mkdir -p /var/opt/simplex /etc/opt/simplex
   sudo chown smp:smp /var/opt/simplex /etc/opt/simplex
   ```

3. Allow `smp-server` port in firewall:

   ```sh
   # For Ubuntu
   ufw allow 5233
   ```

4. **Optional** — If you're using distribution with `systemd`, create `/etc/systemd/system/smp-server.service` file with the following content:

   ```sh
   [Unit]
   Description=SMP server
   [Service]
   User=smp
   Group=smp
   Type=simple
   ExecStart=smp-server start
   ExecStopPost=/usr/bin/env sh -c '[ -e "/var/opt/simplex/smp-server-store.log" ] && cp "/var/opt/simplex/smp-server-store.log" "/var/opt/simplex/smp-server-store.log.bak"'
   KillSignal=SIGINT
   TimeoutStopSec=infinity
   Restart=always
   RestartSec=10
   LimitNOFILE=65535
   [Install]
   WantedBy=multi-user.target
   ```

   And execute `sudo systemctl daemon-reload`.

> ## Configuration

Author: this is a bit confusing - this title should describe INI file after the server is already installed and running, also it seems to be lacking the explanations how to download, when to put it etc.

Also this document should include the references to our update and restart scripts (we can add them to simplexmq repo).

> To investigate which options are available, execute `smp-server` without flags:

Author: ```suggestion
To see which options are available, execute `smp-server` without flags:
```

```sh
sudo su smp -c smp-server

...
Available commands:
  init                     Initialize server - creates /etc/opt/simplex and
                           /var/opt/simplex directories and configuration files
  start                    Start server (configuration:
                           /etc/opt/simplex/smp-server.ini)
  delete                   Delete configuration and log files
```

You can get further help by executing `sudo su smp -c "smp-server <command> -h"`

After that, we need to configure `smp-server`:

### Interactively

Execute the following command:

```sh
sudo su smp -c "smp-server init"
```

> There will be several options to consider:

Author: ```suggestion
There are several options to consider:
```

- `Enable store log to restore queues and messages on server restart (Yn):`

>   Enter `y` to enable saving and restoring messages upon restarting the server.

Author: ```suggestion
  Enter `y` to enable saving and restoring connections and messages when the server is restarted.
  
  _Please note_: it is important to use SIGINT to restart the server, as otherwise the undelivered messages will not be restored. The connections will be restored irrespective of how the server is restarted, as unlike messages they are added to append-only log on every change.
```

  
- `Enable logging daily statistics (yN):`

>   Enter `y` to enable logging statistics for `Prometheus/Grafana`.

Author: ```suggestion
  Enter `y` to enable logging statistics in CSV format, e.g. they can be used to show aggregate usage charts in `Grafana`.
  
These statistics include daily counts of created, secured and deleted queues, sent and received messages, and also daily, weekly, and monthly counts of active queues (that is, the queues that were used for any messages). This information does not include anything that would allow correlating different queues as belonging to the same users.
```

  
- `Require a password to create new messaging queues?`

  Enter `r` or your arbitrary password to password-protect `smp-server`, or `n` to disable password protection.
  
- `Enter server FQDN or IP address for certificate (127.0.0.1):`

>   Enter your domain or ip address to serve `smp-server` from.

Author: ```suggestion
Enter your domain or ip address that your smp-server is running on - it will be included in server certificates and also printed as part of server address.
```

> ### Manually

Author: ```suggestion
### Via command line options
```

Execute the following command:

```sh
sudo su smp -c "smp-server init -h"

...
Available options:
  -l,--store-log           Enable store log for persistence
  -s,--daily-stats         Enable logging daily server statistics
  -a,--sign-algorithm ALG  Signature algorithm used for TLS certificates:
                           ED25519, ED448 (default: ED448)
  --ip IP                  Server IP address, used as Common Name for TLS online
                           certificate if FQDN is not supplied
                           (default: "127.0.0.1")
  -n,--fqdn FQDN           Server FQDN used as Common Name for TLS online
                           certificate
  --no-password            Allow creating new queues without password
  --password PASSWORD      Set password to create new messaging queues
  -y,--yes                 Non-interactive initialization using command-line
                           options
  -h,--help                Show this help text
```

> You should determine which flags are valid for your use-case and then execute `smp-server init` with `-y` flag for non-interactive initialization:

Author: ```suggestion
You should determine which flags are needed for your use-case and then execute `smp-server init` with `-y` flag for non-interactive initialization:
```

```sh
sudo su smp -c "smp-server init -y -<your flag> <your option>"
```

> For example:

Author: ```suggestion
For example, run:
```

```sh
sudo su smp -c "smp-server init -y -l --ip 192.168.1.5 --password test"
```

> ... to initilize your `smp-server` configuration with `-l` flag to enable storing messages, `--ip` flag with `192.168.1.5` value to configure using ip and with `--password` flag with `test` value to password-protect `smp-server`.

Author: ```suggestion
to initilize your `smp-server` configuration with `-l` flag to enable storing messages, `--ip` flag with `192.168.1.5` value to configure using ip and with `--password` flag with `test` value to password-protect `smp-server`.
```

---

After that, your installation is complete and you should see in your teminal output something like this:

```sh
Certificate request self-signature ok
subject=CN = 127.0.0.1
Server initialized, you can modify configuration in /etc/opt/simplex/smp-server.ini.
Run `smp-server start` to start server.
----------
You should store CA private key securely and delete it from the server.
If server TLS credential is compromised this key can be used to sign a new one, keeping the same server identity and established connections.
CA private key location: /etc/opt/simplex/ca.key
----------
SMP server v3.4.0
Fingerprint: d5fcsc7hhtPpexYUbI2XPxDbyU2d3WsVmROimcL90ss=
Server address: smp://d5fcsc7hhtPpexYUbI2XPxDbyU2d3WsVmROimcL90ss=:V8ONoJ6ICwnrZnTC_QuSHfCEYq53uLaJKQ_oIC6-ve8=@<hostnames>
```

> 

Author: ```suggestion

The server address above should be used in your client configuration and if you added server password it should only be shared with the other people when you want to allow them to use your server to receive the messages (all your contacts will be able to send messages, as it does not require a password). If you passed IP address or hostnames during the initialisation, they will be printed as part of server address, otherwise replace `<hostnames>` with the actual server addresses.
```

## Documentation

All necessary files for `smp-server` are located in `/etc/opt/simplex/` folder.

> Stored messages and message queues are located in `/var/opt/simplex/` folder.

Author: ```suggestion
Stored messages, connections, statistics and server log are located in `/var/opt/simplex/` folder.
```

> ### smp link

Author: ```suggestion
### SMP server address
```
